### PR TITLE
break(aws-amplify-angular): service modularization 

### PR DESCRIPTION
**Pull request by haverchuck**
*Description of changes:*
This PR consists of:

- PRIMARY CHANGE:  Changes to the AmplifyService so that customers can pass in the specific Amplify modules that they want to use, with the goal of reducing bundle size:
```
  providers: [
    {
      provide: AmplifyService,
      useFactory:  () => {
        return AmplifyModules({
          Auth,
          Storage,
          Interactions
        });
      }
    }
  ],
```
This requires us to NOT use classes in the AmplifyService, as we are trying to avoid importing Amplify modules within this service file.

ADDITIONAL CHANGES:

- We have added many unit tests, and have added the @Inject directive in the components to support unit testing with the new service.  

- Changed AmplifyService to protected (customer suggestion).

- Linting, linting, linting...

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/aws-amplify-angular/src/components/xr/sumerian-scene-component/sumerian-scene.component.core.ts [hunk 1]
```diff
@@ -99,18 +96,17 @@ export class SumerianSceneComponentCore implements OnInit, OnDestroy {
...
+  protected logger: any;
 
   @Input()
   set data(data: any) {
     this.sceneName = data.sceneName;
   }
 
-  constructor(amplifyService: AmplifyService) {
-    this.amplifyService = amplifyService;
+  constructor(protected amplifyService: AmplifyService) {
```
Comment (elorzafe) [inline packages/aws-amplify-angular/src/components/xr/sumerian-scene-component/sumerian-scene.component.core.ts]: is this protected doing an implicit `protected amplifyService`?

File: packages/aws-amplify-angular/src/components/authenticator/authenticator/authenticator.component.core.ts [hunk 2]
```diff
@@ -61,19 +61,30 @@ const template = `
...
-  amplifyService: AmplifyService;
-
-  constructor(amplifyService: AmplifyService) {
-    this.amplifyService = amplifyService;
+  constructor(protected amplifyService: AmplifyService) {
     this.subscribe();
   }
 
+  ngOnInit() {
+    const loadStatus = this.amplifyService.auth().currentAuthenticatedUser()
```
Comment (elorzafe) [inline packages/aws-amplify-angular/src/components/authenticator/authenticator/authenticator.component.core.ts]: You need to check before if `this.amplifyService.auth()` is not `undefined`?

File: packages/aws-amplify-angular/src/components/authenticator/confirm-sign-in-component/confirm-sign-in-component.core.ts [hunk 3]
```diff
@@ -33,48 +35,44 @@ const template = `
...
   _show: boolean;
   code: string;
   errorMessage: string;
-  amplifyService: AmplifyService;
+  protected logger: any;
 
-  constructor(amplifyService: AmplifyService) {
-    this.amplifyService = amplifyService;
+  constructor(@Inject(AmplifyService) protected amplifyService: AmplifyService) {
+    this.logger = this.amplifyService.logger('ConfiSignInComponent');
```
Comment (elorzafe) [inline packages/aws-amplify-angular/src/components/authenticator/confirm-sign-in-component/confirm-sign-in-component.core.ts]: could be `this.amplifyService.logger` `undefined`?

File: packages/aws-amplify-angular/src/components/authenticator/confirm-sign-in-component/confirm-sign-in-component.ionic.ts [hunk 4]
```diff
@@ -1,14 +1,19 @@
...
+import { AmplifyService } from '../../../providers/amplify.service';
+import { AuthState } from '../../../providers/auth.state';
+import { ConfirmSignInComponentCore } from './confirm-sign-in-component.core';
 
 const template = `
 <div class="amplify-form-container" *ngIf="_show">
   <div class="amplify-form-body">
-    <div class="amplify-form-header amplify-form-header-ionic">{{ this.amplifyService.i18n().get('Confirm your sign in code') }}</div>
+    <div class="amplify-form-header amplify-form-header-ionic">
+      {{ this.amplifyService.i18n().get('Confirm your sign in code') }}
```
Comment (elorzafe) [inline packages/aws-amplify-angular/src/components/authenticator/confirm-sign-in-component/confirm-sign-in-component.ionic.ts]: could be `this.amplifyService.i18n` `undefined`?

File: packages/aws-amplify-angular/src/components/authenticator/confirm-sign-up-component/confirm-sign-up.component.core.ts [hunk 5]
```diff
@@ -22,9 +22,10 @@ const template = `
 <div class="amplify-form-container">
   <div class="amplify-form-body">
     <div class="amplify-form-header">{{ this.amplifyService.i18n().get('Confirm Sign up') }}</div>
-
     <div class="amplify-form-row">
-      <label class="amplify-input-label" for="amplifyUsername"> {{ this.amplifyService.i18n().get('Username *') }}</label>
+      <label class="amplify-input-label" for="amplifyUsername">
+        {{ this.amplifyService.i18n().get('Username *') }}
```
Comment (elorzafe) [inline packages/aws-amplify-angular/src/components/authenticator/confirm-sign-up-component/confirm-sign-up.component.core.ts]: could be `this.amplifyService.i18n` `undefined`?

File: packages/aws-amplify-angular/src/components/xr/sumerian-scene-component/sumerian-scene.component.core.ts [hunk 6]
```diff
@@ -99,18 +96,17 @@ export class SumerianSceneComponentCore implements OnInit, OnDestroy {
...
     this.amplifyUI = AmplifyUI;
+    this.logger = this.amplifyService.logger('SumerianSceneComponentCore');
   }
 
   ngOnInit() {
```
Comment (elorzafe) [inline packages/aws-amplify-angular/src/components/xr/sumerian-scene-component/sumerian-scene.component.core.ts]: You need to check if `this.amplifyService.xr` is not undefined.

File: packages/aws-amplify-angular/src/providers/amplify.service.ts [hunk 7]
```diff
@@ -1,62 +1,75 @@
...
+  import { Injectable, Optional, Inject } from '@angular/core';
+  import { Subject } from 'rxjs/Subject';
+  import { Observable } from 'rxjs/Observable';
+  import Amplify, { Logger, I18n } from '@aws-amplify/core';
+  import { AuthState } from './auth.state';
+  import { authDecorator } from './auth.decorator';
+  
+  @Injectable()
+  export class AmplifyService {
+    private _auth: any;
```
[the remark this answers is not in the document]
Comment (haverchuck) [reply packages/aws-amplify-angular/src/providers/amplify.service.ts]: Not possible at this time without breaking modularization.

File: packages/aws-amplify-angular/src/providers/auth.decorator.ts [hunk 8]
```diff
@@ -21,28 +21,26 @@ function check(authState: Subject<AuthState>) {
 function listen(authState: Subject<AuthState>) {
   const config = Amplify.configure(null);
   if (_.has(config, 'Auth.oauth')) {
-    Hub.listen(
-      'auth',
-      capsule => {
+    Hub.listen('auth', {
+      onHubCapsule: capsule => {
```
[the remark this answers is not in the document]
Comment (haverchuck) [reply packages/aws-amplify-angular/src/providers/auth.decorator.ts]: hmm.... gonna need to address that.

File: packages/aws-amplify-angular/src/components/authenticator/confirm-sign-in-component/confirm-sign-in-component.core.ts (hunk 3 above)
Comment (haverchuck) [reply packages/aws-amplify-angular/src/components/authenticator/confirm-sign-in-component/confirm-sign-in-component.core.ts]: I think Logger would be present by default

File: packages/aws-amplify-angular/src/components/authenticator/authenticator/authenticator.component.core.ts (hunk 2 above)
Comment (haverchuck) [reply packages/aws-amplify-angular/src/components/authenticator/authenticator/authenticator.component.core.ts]: Done.

File: packages/aws-amplify-angular/src/components/xr/sumerian-scene-component/sumerian-scene.component.core.ts (hunk 6 above)
Comment (haverchuck) [reply packages/aws-amplify-angular/src/components/xr/sumerian-scene-component/sumerian-scene.component.core.ts]: Done

File: packages/aws-amplify-angular/src/components/authenticator/confirm-sign-up-component/confirm-sign-up.component.core.ts (hunk 5 above)
Comment (haverchuck) [reply packages/aws-amplify-angular/src/components/authenticator/confirm-sign-up-component/confirm-sign-up.component.core.ts]: i18n should be present by default.

File: packages/aws-amplify-angular/src/components/authenticator/confirm-sign-in-component/confirm-sign-in-component.ionic.ts (hunk 4 above)
Comment (haverchuck) [reply packages/aws-amplify-angular/src/components/authenticator/confirm-sign-in-component/confirm-sign-in-component.ionic.ts]: Should be present by default

File: packages/aws-amplify-angular/src/providers/amplify.service.ts [hunk 9]
```diff
@@ -1,62 +1,75 @@
...
+  *
+  * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+  * CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
+  * and limitations under the License.
+  */
+  // tslint:enable
+
+  import { Injectable, Optional, Inject } from '@angular/core';
+  import { Subject } from 'rxjs/Subject';
+  import { Observable } from 'rxjs/Observable';
```
[the remark this answers is not in the document]
Comment (haverchuck) [reply packages/aws-amplify-angular/src/providers/amplify.service.ts]: Done

File: packages/aws-amplify-angular/package.json [hunk 10]
```diff
@@ -70,6 +69,7 @@
     "aws-amplify": "^1.x"
```
[the remark this answers is not in the document]
Comment (haverchuck) [reply packages/aws-amplify-angular/package.json]: We will throw errors if not present.

File: packages/aws-amplify-angular/src/providers/auth.decorator.ts [hunk 11]
```diff
@@ -128,12 +126,11 @@ function decorateConfirmSignUp(authState: Subject<AuthState>) {
   };
 }
 
-export function authDecorator(authState: Subject<AuthState>) {
-  check(authState);
+export function authDecorator(authState: Subject<AuthState>, authModule) {
```
[the remark this answers is not in the document]
Comment (haverchuck) [reply packages/aws-amplify-angular/src/providers/auth.decorator.ts]: Done

File: packages/aws-amplify-angular/src/providers/amplify.service.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (LosD) [reply packages/aws-amplify-angular/src/providers/amplify.service.ts]: Thanks for breaking typing. Development is _great_ without IDE hints!

**Pull request merged**